### ISSUE TITLE
toJs conversion of dict keys

### ISSUE THREAD

**Issue by hoodmane** [iodide-project/pyodide#1245]
Dictionaries allow keys not to be a string, but in an `Object` all keys are strings. This may cause unexpected data loss:
```javascript
>>> pyodide.runPython(`{ 1 : 7, "1" : 9}`).toJs()
{1: 9}
>>> pyodide.runPython(`{ "1" : 9, 1 : 7}`).toJs()
{1: 7}
```
We could use a `Map` instead of an `Object` to fix this problem. 

Currently using a tuple key doesn't lead to the best result either:
```javascript
>>> pyodide.runPython(`{ (2,3) : 6}`).toJs()
{ "2,3" : 6 }
```
If the key is a `bytes object this is weird too:
```javascript
>>> pyodide.runPython(`{ b"abc" : 6}`).toJs()
{"97,98,99": 6}
```
One thing to consider is at least to use `str` on the key rather than doing `key.toJs().toString()` like we are doing now. Then at least we'd convert `{ (2,3) : 6}` to `{ "(2, 3)" : 6}`. The bytes example would give `{ "b'abc'" : 6}` which isn't great but is probably better than the current behavior.

One annoying thing about `Map` is that it compares on memory address:
```javascript
>>> let m = new Map();
>>> m.set([2,3], 7);
>>> m.get([2,3]); 
undefined // second [2,3] is not at the same memory address as the first
```
For this reason it might still be best to use `str` on the keys even when returning a `Map` except when integers are used as keys.

**Comment by rth** [comment]
> For this reason it might still be best to use str on the keys even when returning a Map except when integers are used as keys.

Thanks for the analysis! That sounds very reasonable to me given the `Map` constraints. This should be a warning in the doc somewhere though.

**Comment by hoodmane** [comment]
The other most reasonable option is to throw an error, maybe that would be better actually.

**Comment by rth** [comment]
+1 for error in that case.

**Comment by hoodmane** [comment]
Okay, I'll open a PR for this.

**Comment by hoodmane** [comment]
Closed by #1248.
